File: api/vana.py
"""
Mera Vana (मेरा वन - My Garden) API Endpoint.

Plant collection CRUD operations.
Requires authentication for all operations.
"""

import json
import logging
from http.server import BaseHTTPRequestHandler
from typing import Dict, Any
from urllib.parse import urlparse, parse_qs

from _utils.database import get_database
from _utils.auth_middleware import verify_token

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def send_response(handler: BaseHTTPRequestHandler, status: int, data: Dict[str, Any]) -> None:
    """Send JSON response with proper headers."""
    handler.send_response(status)
    handler.send_header('Content-Type', 'application/json')
    handler.send_header('Access-Control-Allow-Origin', '*')
    handler.send_header('Access-Control-Allow-Methods', 'GET, POST, PATCH, DELETE, OPTIONS')
    handler.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class handler(BaseHTTPRequestHandler):
    """Vercel serverless handler for Mera Vana endpoint."""

    def do_OPTIONS(self):
        """Handle CORS preflight requests."""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PATCH, DELETE, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.end_headers()

    def do_GET(self):
        """
        Get all plants in user's Mera Vana collection.

        Request headers:
            Authorization: Bearer <token>

        Response:
            {
                "success": true,
                "vana": [...],
                "count": 5
            }
        """
        try:
            # Verify authentication
            try:
                auth_data = verify_token(self)
                user_id = auth_data['user_id']
            except Exception as e:
                logger.warning(f"Authentication failed: {str(e)}")
                send_response(self, 401, {
                    'success': False,
                    'error': str(e)
                })
                return

            # Get plants from database
            db = get_database()
            plants = db.get_mera_vana(user_id)

            send_response(self, 200, {
                'success': True,
                'vana': plants,
                'count': len(plants)
            })

            logger.info(f"Retrieved {len(plants)} plants for user {user_id}")

        except Exception as e:
            logger.error(f"Failed to fetch Mera Vana: {str(e)}")
            send_response(self, 500, {
                'success': False,
                'error': 'Failed to fetch plants. Please try again.'
            })

    def do_POST(self):
        """
        Add a new plant to Mera Vana.

        Request headers:
            Authorization: Bearer <token>

        Request body:
            {
                "name": "My Monstera",
                "species": "Monstera Deliciosa",
                "scientific_name": "Monstera deliciosa",
                "sanskrit_name": null,
                "family": "Araceae",
                "image_url": "https://...",
                "care_schedule": {...},
                "notes": "Birthday gift from Mom"
            }

        Response:
            {
                "success": true,
                "plant_id": "uuid",
                "message": "Plant added to Mera Vana successfully"
            }
        """
        try:
            # Verify authentication
            try:
                auth_data = verify_token(self)
                user_id = auth_data['user_id']
            except Exception as e:
                logger.warning(f"Authentication failed: {str(e)}")
                send_response(self, 401, {
                    'success': False,
                    'error': str(e)
                })
                return

            # Parse request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            plant_data = json.loads(body)

            # Validate required fields
            if not plant_data.get('name'):
                send_response(self, 400, {
                    'success': False,
                    'error': 'Plant name is required'
                })
                return

            if not plant_data.get('species'):
                send_response(self, 400, {
                    'success': False,
                    'error': 'Plant species is required'
                })
                return

            # Save to database
            db = get_database()
            plant_id = db.save_to_vana(user_id, plant_data)

            send_response(self, 200, {
                'success': True,
                'plant_id': plant_id,
                'message': 'Plant added to Mera Vana successfully'
            })

            logger.info(f"Plant {plant_id} added to Vana for user {user_id}")

        except json.JSONDecodeError:
            logger.error("Invalid JSON in request body")
            send_response(self, 400, {
                'success': False,
                'error': 'Invalid JSON in request body'
            })

        except Exception as e:
            logger.error(f"Failed to add plant: {str(e)}")
            send_response(self, 500, {
                'success': False,
                'error': 'Failed to add plant. Please try again.'
            })

    def do_PATCH(self):
        """
        Update a plant in Mera Vana.

        Request headers:
            Authorization: Bearer <token>

        Request body:
            {
                "plant_id": "uuid",
                "last_watered": "2025-11-01T10:30:00Z",
                "watering_frequency_days": 7,
                "notes": "Looking healthier!",
                "health_status": "healthy"
            }

        Response:
            {
                "success": true,
                "message": "Plant updated successfully"
            }
        """
        try:
            # Verify authentication
            try:
                auth_data = verify_token(self)
                user_id = auth_data['user_id']
            except Exception as e:
                logger.warning(f"Authentication failed: {str(e)}")
                send_response(self, 401, {
                    'success': False,
                    'error': str(e)
                })
                return

            # Parse request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            data = json.loads(body)

            plant_id = data.get('plant_id')
            if not plant_id:
                send_response(self, 400, {
                    'success': False,
                    'error': 'plant_id is required'
                })
                return

            # Remove plant_id from updates
            updates = {k: v for k, v in data.items() if k != 'plant_id'}

            if not updates:
                send_response(self, 400, {
                    'success': False,
                    'error': 'No fields to update'
                })
                return

            # Update in database
            db = get_database()
            db.update_vana_plant(plant_id, user_id, updates)

            send_response(self, 200, {
                'success': True,
                'message': 'Plant updated successfully'
            })

            logger.info(f"Plant {plant_id} updated by user {user_id}")

        except json.JSONDecodeError:
            logger.error("Invalid JSON in request body")
            send_response(self, 400, {
                'success': False,
                'error': 'Invalid JSON in request body'
            })

        except Exception as e:
            logger.error(f"Failed to update plant: {str(e)}")
            send_response(self, 500, {
                'success': False,
                'error': str(e)
            })

    def do_DELETE(self):
        """
        Delete a plant from Mera Vana.

        Request headers:
            Authorization: Bearer <token>

        Query parameters:
            plant_id: UUID of the plant to delete

        Response:
            {
                "success": true,
                "message": "Plant removed from Mera Vana successfully"
            }
        """
        try:
            # Verify authentication
            try:
                auth_data = verify_token(self)
                user_id = auth_data['user_id']
            except Exception as e:
                logger.warning(f"Authentication failed: {str(e)}")
                send_response(self, 401, {
                    'success': False,
                    'error': str(e)
                })
                return

            # Extract plant_id from query parameters
            parsed_url = urlparse(self.path)
            query_params = parse_qs(parsed_url.query)
            plant_id = query_params.get('plant_id', [None])[0]

            if not plant_id:
                send_response(self, 400, {
                    'success': False,
                    'error': 'plant_id query parameter is required'
                })
                return

            # Delete from database
            db = get_database()
            db.remove_from_vana(plant_id, user_id)

            send_response(self, 200, {
                'success': True,
                'message': 'Plant removed from Mera Vana successfully'
            })

            logger.info(f"Plant {plant_id} removed by user {user_id}")

        except Exception as e:
            logger.error(f"Failed to delete plant: {str(e)}")
            send_response(self, 500, {
                'success': False,
                'error': str(e)
            })

File: api/vana.js
/**
 * Mera Vana (मेरा वन - My Garden) API Endpoint - Node.js Serverless
 *
 * Plant collection CRUD operations.
 * Requires authentication for all operations.
 */

import { getDatabase } from './_utils/database.js';
import { verifyToken } from './_utils/auth_middleware.js';

/**
 * Send JSON response with proper CORS headers.
 */
function sendResponse(res, status, data) {
  res.status(status).json(data);
}

/**
 * Handle CORS preflight requests.
 */
function handleCORS(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return true;
  }
  return false;
}

/**
 * Authenticate request and return user_id.
 *
 * @param {Object} req - HTTP request object
 * @param {Object} res - HTTP response object
 * @returns {string|null} user_id if authenticated, null if failed (response already sent)
 */
function authenticate(req, res) {
  try {
    const authData = verifyToken(req);
    return authData.user_id;
  } catch (error) {
    console.error('Authentication failed:', error.message);
    sendResponse(res, 401, {
      success: false,
      error: error.message
    });
    return null;
  }
}

/**
 * Extract plant ID from URL path.
 *
 * @param {string} path - URL path
 * @returns {string|null} plant_id or null
 */
function extractPlantId(path) {
  // Match pattern: /api/vana/{plant_id}
  const match = path.match(/\/api\/vana\/([a-f0-9-]+)/i);
  return match ? match[1] : null;
}

/**
 * Handle GET requests - List plants or get specific plant.
 *
 * GET /api/vana - List all user's plants
 * GET /api/vana/:id - Get specific plant
 *
 * Response:
 *   {
 *     "success": true,
 *     "vana": [...],
 *     "count": 5
 *   }
 */
async function handleGet(req, res) {
  try {
    // Authenticate
    const userId = authenticate(req, res);
    if (!userId) return;

    const plantId = extractPlantId(req.url);

    // Get database connection
    const db = getDatabase();

    if (plantId) {
      // Get specific plant
      const plants = await db.getMeraVana(userId);
      const plant = plants.find(p => p.id === plantId);

      if (!plant) {
        return sendResponse(res, 404, {
          success: false,
          error: 'Plant not found'
        });
      }

      sendResponse(res, 200, {
        success: true,
        plant: plant
      });

      console.log(`Retrieved plant ${plantId} for user ${userId}`);
    } else {
      // Get all plants
      const plants = await db.getMeraVana(userId);

      sendResponse(res, 200, {
        success: true,
        vana: plants,
        count: plants.length
      });

      console.log(`Retrieved ${plants.length} plants for user ${userId}`);
    }
  } catch (error) {
    console.error('Failed to fetch Mera Vana:', error.message);
    sendResponse(res, 500, {
      success: false,
      error: 'Failed to fetch plants. Please try again.'
    });
  }
}

/**
 * Handle POST requests - Add new plant.
 *
 * Request body:
 *   {
 *     "name": "My Monstera",
 *     "species": "Monstera Deliciosa",
 *     "scientific_name": "Monstera deliciosa",
 *     "sanskrit_name": null,
 *     "family": "Araceae",
 *     "image_url": "https://...",
 *     "care_schedule": {...},
 *     "notes": "Birthday gift from Mom"
 *   }
 *
 * Response:
 *   {
 *     "success": true,
 *     "plant_id": "uuid",
 *     "message": "Plant added to Mera Vana successfully"
 *   }
 */
async function handlePost(req, res) {
  try {
    // Authenticate
    const userId = authenticate(req, res);
    if (!userId) return;

    const plantData = req.body;

    // Validate required fields
    if (!plantData.name) {
      return sendResponse(res, 400, {
        success: false,
        error: 'Plant name is required'
      });
    }

    if (!plantData.species) {
      return sendResponse(res, 400, {
        success: false,
        error: 'Plant species is required'
      });
    }

    // Get database connection
    const db = getDatabase();

    // Save to database
    const plantId = await db.saveToVana(userId, plantData);

    sendResponse(res, 200, {
      success: true,
      plant_id: plantId,
      message: 'Plant added to Mera Vana successfully'
    });

    console.log(`Plant ${plantId} added to Vana for user ${userId}`);
  } catch (error) {
    console.error('Failed to add plant:', error.message);
    sendResponse(res, 500, {
      success: false,
      error: 'Failed to add plant. Please try again.'
    });
  }
}

/**
 * Handle PUT requests - Update plant.
 *
 * Request body:
 *   {
 *     "plant_id": "uuid",  // Can also be in URL path
 *     "last_watered": "2025-11-01T10:30:00Z",
 *     "watering_frequency_days": 7,
 *     "notes": "Looking healthier!",
 *     "health_status": "healthy"
 *   }
 *
 * Response:
 *   {
 *     "success": true,
 *     "message": "Plant updated successfully"
 *   }
 */
async function handlePut(req, res) {
  try {
    // Authenticate
    const userId = authenticate(req, res);
    if (!userId) return;

    const data = req.body;

    // Get plant_id from URL or body
    let plantId = extractPlantId(req.url) || data.plant_id;

    if (!plantId) {
      return sendResponse(res, 400, {
        success: false,
        error: 'plant_id is required'
      });
    }

    // Remove plant_id from updates
    const updates = { ...data };
    delete updates.plant_id;

    if (Object.keys(updates).length === 0) {
      return sendResponse(res, 400, {
        success: false,
        error: 'No fields to update'
      });
    }

    // Get database connection
    const db = getDatabase();

    // Update in database
    await db.updateVanaPlant(plantId, userId, updates);

    sendResponse(res, 200, {
      success: true,
      message: 'Plant updated successfully'
    });

    console.log(`Plant ${plantId} updated by user ${userId}`);
  } catch (error) {
    console.error('Failed to update plant:', error.message);
    sendResponse(res, 500, {
      success: false,
      error: error.message
    });
  }
}

/**
 * Handle DELETE requests - Remove plant.
 *
 * URL: /api/vana/:id or /api/vana?plant_id=uuid
 *
 * Response:
 *   {
 *     "success": true,
 *     "message": "Plant removed from Mera Vana successfully"
 *   }
 */
async function handleDelete(req, res) {
  try {
    // Authenticate
    const userId = authenticate(req, res);
    if (!userId) return;

    // Get plant_id from URL path or query parameter
    let plantId = extractPlantId(req.url);

    if (!plantId && req.query && req.query.plant_id) {
      plantId = req.query.plant_id;
    }

    if (!plantId) {
      return sendResponse(res, 400, {
        success: false,
        error: 'plant_id is required in URL path or query parameter'
      });
    }

    // Get database connection
    const db = getDatabase();

    // Delete from database
    await db.removeFromVana(plantId, userId);

    sendResponse(res, 200, {
      success: true,
      message: 'Plant removed from Mera Vana successfully'
    });

    console.log(`Plant ${plantId} removed by user ${userId}`);
  } catch (error) {
    console.error('Failed to delete plant:', error.message);
    sendResponse(res, 500, {
      success: false,
      error: error.message
    });
  }
}

/**
 * Main handler function for Vercel serverless deployment.
 *
 * Routes:
 *   GET /api/vana - List user's plants
 *   POST /api/vana - Add new plant
 *   GET /api/vana/:id - Get specific plant
 *   PUT /api/vana/:id - Update plant
 *   DELETE /api/vana/:id - Delete plant
 */
export default async function handler(req, res) {
  // Handle CORS
  if (handleCORS(req, res)) {
    return;
  }

  // Route based on HTTP method
  switch (req.method) {
    case 'GET':
      return handleGet(req, res);
    case 'POST':
      return handlePost(req, res);
    case 'PUT':
    case 'PATCH':
      return handlePut(req, res);
    case 'DELETE':
      return handleDelete(req, res);
    default:
      return sendResponse(res, 405, {
        success: false,
        error: 'Method not allowed'
      });
  }
}
